<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#EDA" data-toc-modified-id="EDA-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>EDA</a></span></li><li><span><a href="#Pisah-data" data-toc-modified-id="Pisah-data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Pisah data</a></span></li><li><span><a href="#Kualitas-Model" data-toc-modified-id="Kualitas-Model-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Kualitas Model</a></span><ul class="toc-item"><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span><ul class="toc-item"><li><span><a href="#Tanpa-Tuning-Hyperparameter" data-toc-modified-id="Tanpa-Tuning-Hyperparameter-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Tanpa Tuning Hyperparameter</a></span></li><li><span><a href="#Tuning-Hyperparameter" data-toc-modified-id="Tuning-Hyperparameter-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Tuning Hyperparameter</a></span><ul class="toc-item"><li><span><a href="#Max-Depth" data-toc-modified-id="Max-Depth-3.1.2.1"><span class="toc-item-num">3.1.2.1&nbsp;&nbsp;</span>Max Depth</a></span></li><li><span><a href="#Criterion" data-toc-modified-id="Criterion-3.1.2.2"><span class="toc-item-num">3.1.2.2&nbsp;&nbsp;</span>Criterion</a></span></li><li><span><a href="#Splitter" data-toc-modified-id="Splitter-3.1.2.3"><span class="toc-item-num">3.1.2.3&nbsp;&nbsp;</span>Splitter</a></span></li><li><span><a href="#max_features" data-toc-modified-id="max_features-3.1.2.4"><span class="toc-item-num">3.1.2.4&nbsp;&nbsp;</span>max_features</a></span></li></ul></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Kesimpulan</a></span></li></ul></li><li><span><a href="#RandomForestClassifier" data-toc-modified-id="RandomForestClassifier-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>RandomForestClassifier</a></span><ul class="toc-item"><li><span><a href="#Tanpa-Tuning-Hyperparameter" data-toc-modified-id="Tanpa-Tuning-Hyperparameter-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Tanpa Tuning Hyperparameter</a></span></li><li><span><a href="#Tuning-Hyperparameter" data-toc-modified-id="Tuning-Hyperparameter-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Tuning Hyperparameter</a></span><ul class="toc-item"><li><span><a href="#n_estimators" data-toc-modified-id="n_estimators-3.2.2.1"><span class="toc-item-num">3.2.2.1&nbsp;&nbsp;</span>n_estimators</a></span></li><li><span><a href="#criterion" data-toc-modified-id="criterion-3.2.2.2"><span class="toc-item-num">3.2.2.2&nbsp;&nbsp;</span>criterion</a></span></li><li><span><a href="#Max_depth" data-toc-modified-id="Max_depth-3.2.2.3"><span class="toc-item-num">3.2.2.3&nbsp;&nbsp;</span>Max_depth</a></span></li><li><span><a href="#max_features" data-toc-modified-id="max_features-3.2.2.4"><span class="toc-item-num">3.2.2.4&nbsp;&nbsp;</span>max_features</a></span></li><li><span><a href="#bootstrap" data-toc-modified-id="bootstrap-3.2.2.5"><span class="toc-item-num">3.2.2.5&nbsp;&nbsp;</span>bootstrap</a></span></li><li><span><a href="#warm_start" data-toc-modified-id="warm_start-3.2.2.6"><span class="toc-item-num">3.2.2.6&nbsp;&nbsp;</span>warm_start</a></span></li><li><span><a href="#class_weight" data-toc-modified-id="class_weight-3.2.2.7"><span class="toc-item-num">3.2.2.7&nbsp;&nbsp;</span>class_weight</a></span></li></ul></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Kesimpulan</a></span></li></ul></li><li><span><a href="#Logistic-Regression" data-toc-modified-id="Logistic-Regression-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Logistic Regression</a></span><ul class="toc-item"><li><span><a href="#Tanpa-Tuning-Hyperparameter" data-toc-modified-id="Tanpa-Tuning-Hyperparameter-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Tanpa Tuning Hyperparameter</a></span></li><li><span><a href="#Tuning-Hyperparameter" data-toc-modified-id="Tuning-Hyperparameter-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Tuning Hyperparameter</a></span><ul class="toc-item"><li><span><a href="#solver" data-toc-modified-id="solver-3.3.2.1"><span class="toc-item-num">3.3.2.1&nbsp;&nbsp;</span>solver</a></span></li><li><span><a href="#penalty" data-toc-modified-id="penalty-3.3.2.2"><span class="toc-item-num">3.3.2.2&nbsp;&nbsp;</span>penalty</a></span></li><li><span><a href="#multi_class" data-toc-modified-id="multi_class-3.3.2.3"><span class="toc-item-num">3.3.2.3&nbsp;&nbsp;</span>multi_class</a></span></li><li><span><a href="#warm_start" data-toc-modified-id="warm_start-3.3.2.4"><span class="toc-item-num">3.3.2.4&nbsp;&nbsp;</span>warm_start</a></span></li></ul></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Kesimpulan</a></span></li></ul></li></ul></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Kesimpulan</a></span></li></ul></div>

# Deskripsi Project
Mengembangkan sebuah model yang dapat menganalisis perilaku konsumen dan merekomendasikan salah satu dari kedua paket terbaru Megaline: Smart atau Ultra. menggunakan Machine Learning Supervised dengan 3 model: **DecisionTreeClassifier, RandomForestClassifier dan LogisticRegresiion**.
Sekaligus Mengamati perbedaan antara model yang **Tanpa Tuning Hyperparameter** dengan model yang menggunakan **Tuning Hyperparameter** sampai mendapatkan Tingkat akurasi yang lebih tinggi dari 75%


## EDA

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [ ]:
df = pd.read_csv('/datasets/users_behavior.csv')

In [ ]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [ ]:
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [ ]:
df.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [ ]:
df.duplicated().sum()

0

Sanity Check

In [ ]:
df['is_ultra'].value_counts()/len(df['is_ultra']) * 100

0    69.352831
1    30.647169
Name: is_ultra, dtype: float64

**KESIMPULAN**

Setelah dilakukan nya EDA pada dataframe **'/datasets/users_behavior.csv'** kita tidak menemukan adanya kesalahan data berupa **Missing Value**, **Duplicated , Kesalahan Tipe Data**. jadi kita bisa langsung melakukan sanity check pada data dan mendapatkan bahwa **Konsumen yang menggunakan Paket Smart sebanyak 69%** sehingga model yang perlu kita buat harus memiliki tingkat akurasi di atas 69%, atau agar lebih baik kita buat tingkat akurasi minimal model adalah 75%

## Pisah data

In [ ]:
features = df.drop(columns='is_ultra',axis=1)
target = df['is_ultra']

In [ ]:
feature_train,feature_valid,target_train,target_valid = train_test_split(features,target,test_size=0.2,random_state=12345)

In [ ]:
feature_train,feature_test, target_train,target_test = train_test_split(feature_train,target_train,
                                                                        test_size=0.25,random_state=12345)

In [ ]:
print('Besaran data untuk Training Set =',feature_train.shape,target_train.shape)
print('Besaran data untuk Validation Set =',feature_valid.shape,target_valid.shape)
print('Besaran data untuk Test Set =',feature_test.shape,target_test.shape)

Besaran data untuk Training Set = (1928, 4) (1928,)
Besaran data untuk Validation Set = (643, 4) (643,)
Besaran data untuk Test Set = (643, 4) (643,)


**KESIMPULAN**

Total data konsumen yang kita punya adalah **3214**, dikarenakan kita hanya memiliki 1 dataframe, oleh karena itu diperlukan **Split/Pembagian** data menjadi data **Training**, data **Validation** dan data **Test** dengan proporsi data sebesar **60:20:20**, 60 untuk training, 20 untuk validation dan 20 untuk test.

Dikarenakan Target kita adalah ingin Merekomendasikan Paket Ultra atau smart , oleh karena itu pisahkan kolom **'is_ultra'** pada dataframe untuk menjadi target/output, dan sisa kolom nya menjadi feature/input.


## Kualitas Model

### DecisionTreeClassifier

#### Tanpa Tuning Hyperparameter

In [ ]:
model_tree = DecisionTreeClassifier(random_state=12345)
model_tree.fit(feature_train,target_train)

predict_tree_valid = model_tree.predict(feature_valid)
result_tree_valid = accuracy_score(target_valid,predict_tree_valid)

predict_tree_test = model_tree.predict(feature_test)
result_tree_test = accuracy_score(target_test,predict_tree_test)

print (f'Hasil akurasi validation = {result_tree_valid*100}, Hasil akurasi test = {result_tree_test*100}')

Hasil akurasi validation = 73.09486780715396, Hasil akurasi test = 71.22861586314151


Hasil dari Model DecisionTreeClassifier tanpa tuning Hyperparameter didapat model mengalami Underfitting/ penurunan hasil.

#### Tuning Hyperparameter

##### Max Depth

In [ ]:

for depth in range(1,20):
    model_tree_1 = DecisionTreeClassifier(random_state=12345,max_depth=depth)
    model_tree_1.fit(feature_train,target_train)

    predict_tree_valid_1 = model_tree_1.predict(feature_valid)
    result_tree_valid_1 = accuracy_score(target_valid,predict_tree_valid_1)

    predict_tree_test_1 = model_tree_1.predict(feature_test)
    result_tree_test_1 = accuracy_score(target_test,predict_tree_test_1)


    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_tree_valid_1*100}, Hasil akurasi test = {result_tree_test_1*100}, max_depth ={depth}')



-----------------------------------------------------------
Hasil akurasi validation = 74.80559875583204, Hasil akurasi test = 73.87247278382581, max_depth =1
-----------------------------------------------------------
Hasil akurasi validation = 78.38258164852256, Hasil akurasi test = 75.73872472783826, max_depth =2
-----------------------------------------------------------
Hasil akurasi validation = 78.69362363919129, Hasil akurasi test = 76.51632970451011, max_depth =3
-----------------------------------------------------------
Hasil akurasi validation = 78.69362363919129, Hasil akurasi test = 76.36080870917574, max_depth =4
-----------------------------------------------------------
Hasil akurasi validation = 78.84914463452566, Hasil akurasi test = 75.89424572317263, max_depth =5
-----------------------------------------------------------
Hasil akurasi validation = 77.91601866251943, Hasil akurasi test = 75.73872472783826, max_depth =6
----------------------------------------------

setelah tuning hyperparameter max_depth dilakukan, max_depth=13 sudah mendapatkan hasil yang sama antara valid dan test, oleh karena itu kita akan mengambil best max_depth=14, karena hasil test lebih besar dari hasil valid.

##### Criterion

In [ ]:
criterion = ['gini','entropy']
for crit in criterion:
    model_tree_2 = DecisionTreeClassifier(random_state=12345,max_depth=14,criterion=crit)
    model_tree_2.fit(feature_train,target_train)

    predict_tree_valid_2 = model_tree_2.predict(feature_valid)
    result_tree_valid_2 = accuracy_score(target_valid,predict_tree_valid_2)

    predict_tree_test_2 = model_tree_2.predict(feature_test)
    result_tree_test_2 = accuracy_score(target_test,predict_tree_test_2)

    print (f'Hasil akurasi validation = {result_tree_valid_2*100}, Hasil akurasi test = {result_tree_test_2*100},criterion = {crit}')

Hasil akurasi validation = 75.42768273716952, Hasil akurasi test = 75.73872472783826,criterion = gini
Hasil akurasi validation = 75.58320373250389, Hasil akurasi test = 75.58320373250389,criterion = entropy


Saya tidak tau kenapa Hyperparameter criterion=log_loss malah menampilkan error, sedangkan gini dan entropy tidak, jika di lihat dari hasil hyperparam criterion= gini /default lebih bagus.

##### Splitter

In [ ]:
splitter = ['best','random']
for split in splitter:
    model_tree_4 = DecisionTreeClassifier(random_state=12345,max_depth=14,criterion='gini',splitter=split)
    model_tree_4.fit(feature_train,target_train)

    predict_tree_valid_4 = model_tree_4.predict(feature_valid)
    result_tree_valid_4 = accuracy_score(target_valid,predict_tree_valid_4)

    predict_tree_test_4 = model_tree_4.predict(feature_test)
    result_tree_test_4 = accuracy_score(target_test,predict_tree_test_4)

    print (f'Hasil akurasi validation = {result_tree_valid_4*100}, Hasil akurasi test = {result_tree_test_4*100},splitter = {split}')

Hasil akurasi validation = 75.42768273716952, Hasil akurasi test = 75.73872472783826,splitter = best
Hasil akurasi validation = 75.73872472783826, Hasil akurasi test = 73.71695178849144,splitter = random


Hyperparameter splitter=best /default sudah cukup bagus menampilkan hasil akurasi test lebih tinggi dari valid

##### max_features

In [ ]:
max_features = ['auto', 'sqrt', 'log2']
for features in max_features:
    model_tree_3 = DecisionTreeClassifier(random_state=12345,max_depth=14,criterion='gini',splitter='best',max_features=features)
    model_tree_3.fit(feature_train,target_train)

    predict_tree_valid_3 = model_tree_3.predict(feature_valid)
    result_tree_valid_3 = accuracy_score(target_valid,predict_tree_valid_3)

    predict_tree_test_3 = model_tree_3.predict(feature_test)
    result_tree_test_3 = accuracy_score(target_test,predict_tree_test_3)

    print (f'Hasil akurasi validation = {result_tree_valid_3*100}, Hasil akurasi test = {result_tree_test_3*100},max_features = {features}')

Hasil akurasi validation = 74.96111975116641, Hasil akurasi test = 75.11664074650078,max_features = auto
Hasil akurasi validation = 74.96111975116641, Hasil akurasi test = 75.11664074650078,max_features = sqrt
Hasil akurasi validation = 74.96111975116641, Hasil akurasi test = 75.11664074650078,max_features = log2


dengan tuning hyperparameter max_feature, membuat hasil valid dan test menurun

#### Kesimpulan

model DecisionTreeClassifier dengan tuning Hyperparameter, random_state=12345, max_depth =14, criterion=gini, splitter=best.
memiliki hasil paling bagus dari yang lain nya

In [ ]:
model_tree_5 = DecisionTreeClassifier(random_state=12345,max_depth=14,criterion='gini',splitter='best')
model_tree_5.fit(feature_train,target_train)

predict_tree_valid_5 = model_tree_5.predict(feature_valid)
result_tree_valid_5 = accuracy_score(target_valid,predict_tree_valid_5)

predict_tree_test_5 = model_tree_5.predict(feature_test)
result_tree_test_5 = accuracy_score(target_test,predict_tree_test_5)

print (f'Hasil akurasi validation = {result_tree_valid_5*100}, Hasil akurasi test = {result_tree_test_5*100}')

Hasil akurasi validation = 75.42768273716952, Hasil akurasi test = 75.73872472783826


### RandomForestClassifier

#### Tanpa Tuning Hyperparameter

In [ ]:
model_forest = RandomForestClassifier(random_state=12345)
model_forest.fit(feature_train,target_train)

predict_forest_valid = model_forest.predict(feature_valid)
result_forest_valid = accuracy_score(target_valid,predict_forest_valid)

predict_forest_test = model_forest.predict(feature_test)
result_forest_test = accuracy_score(target_test,predict_forest_test)

print (f'Hasil akurasi validation = {result_forest_valid*100}, Hasil akurasi test = {result_forest_test*100}')

Hasil akurasi validation = 78.38258164852256, Hasil akurasi test = 79.47122861586314


dari hasil yang didapat tanpa tuning, sudah cukup bagus, model tidak underfitting maupun overfitting, tapi kita coba tuning

#### Tuning Hyperparameter

##### n_estimators

In [ ]:
n_estimators = [ 10,50,100,200,250]
for est in n_estimators:
    model_forest_1 = RandomForestClassifier(random_state=12345,n_estimators=est)
    model_forest_1.fit(feature_train,target_train)

    predict_forest_valid_1 = model_forest_1.predict(feature_valid)
    result_forest_valid_1 = accuracy_score(target_valid,predict_forest_valid_1)

    predict_forest_test_1 = model_forest_1.predict(feature_test)
    result_forest_test_1 = accuracy_score(target_test,predict_forest_test_1)

    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_forest_valid_1*100}, Hasil akurasi test = {result_forest_test_1*100} n_estimators={est} ')

-----------------------------------------------------------
Hasil akurasi validation = 78.69362363919129, Hasil akurasi test = 78.84914463452566 n_estimators=10 
-----------------------------------------------------------
Hasil akurasi validation = 78.84914463452566, Hasil akurasi test = 79.47122861586314 n_estimators=50 
-----------------------------------------------------------
Hasil akurasi validation = 78.38258164852256, Hasil akurasi test = 79.47122861586314 n_estimators=100 
-----------------------------------------------------------
Hasil akurasi validation = 78.69362363919129, Hasil akurasi test = 79.62674961119751 n_estimators=200 
-----------------------------------------------------------
Hasil akurasi validation = 79.00466562986003, Hasil akurasi test = 79.31570762052877 n_estimators=250 


setelah mencoba tuning n_estimator lebih dari 250, hasil test semakin menurun dan kecepatan eksekusi menjadi semakin lambat. jadi hasil terbaik tuning n_estimator adalah 200

##### criterion

In [ ]:
#var criterion sudah ada di model DecissionTreeClassifier
for crit in criterion:
    model_forest_2 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion=crit)
    model_forest_2.fit(feature_train,target_train)

    predict_forest_valid_2 = model_forest_2.predict(feature_valid)
    result_forest_valid_2 = accuracy_score(target_valid,predict_forest_valid_2)

    predict_forest_test_2 = model_forest_2.predict(feature_test)
    result_forest_test_2 = accuracy_score(target_test,predict_forest_test_2)

    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_forest_valid_2*100}, Hasil akurasi test = {result_forest_test_2*100} criterion={crit} ')

-----------------------------------------------------------
Hasil akurasi validation = 78.69362363919129, Hasil akurasi test = 79.62674961119751 criterion=gini 
-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625 criterion=entropy 


tuning entropy membuat hasil valid dan test mengalami peningkatan

##### Max_depth

In [ ]:
for depth in range(1,6):
    model_forest_3 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',max_depth=depth)
    model_forest_3.fit(feature_train,target_train)

    predict_forest_valid_3 = model_forest_3.predict(feature_valid)
    result_forest_valid_3 = accuracy_score(target_valid,predict_forest_valid_3)

    predict_forest_test_3 = model_forest_3.predict(feature_test)
    result_forest_test_3 = accuracy_score(target_test,predict_forest_test_3)

    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_forest_valid_3*100}, Hasil akurasi test = {result_forest_test_3*100} max_depth={depth} ')

-----------------------------------------------------------
Hasil akurasi validation = 75.73872472783826, Hasil akurasi test = 74.33903576982893 max_depth=1 
-----------------------------------------------------------
Hasil akurasi validation = 78.38258164852256, Hasil akurasi test = 76.20528771384136 max_depth=2 
-----------------------------------------------------------
Hasil akurasi validation = 78.53810264385692, Hasil akurasi test = 76.98289269051321 max_depth=3 
-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 77.60497667185071 max_depth=4 
-----------------------------------------------------------
Hasil akurasi validation = 79.16018662519441, Hasil akurasi test = 78.0715396578538 max_depth=5 


jika tuning max_depth pada model random forest membuat hasil test mengalami penurunan, meskipun max_depth diperbesar, jikan dipaksakan malah membuat proses kecepatan program melambat

##### max_features

In [ ]:
for features in max_features:
    model_forest_4 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',max_features=features)
    model_forest_4.fit(feature_train,target_train)

    predict_forest_valid_4 = model_forest_4.predict(feature_valid)
    result_forest_valid_4 = accuracy_score(target_valid,predict_forest_valid_4)

    predict_forest_test_4 = model_forest_4.predict(feature_test)
    result_forest_test_4 = accuracy_score(target_test,predict_forest_test_4)

    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_forest_valid_4*100}, Hasil akurasi test = {result_forest_test_4*100} max_features={features} ')

-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625 max_features=auto 
-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625 max_features=sqrt 
-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625 max_features=log2 


Tidak ada perubahan yang terjadi setelah tuning max_feature, karna tidak ada perubahan selanjutnya akan dibuat default

##### bootstrap

In [ ]:
#dikarenakan default bootstrap adalah True,
model_forest_5 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',bootstrap=False)
model_forest_5.fit(feature_train,target_train)

predict_forest_valid_5 = model_forest_5.predict(feature_valid)
result_forest_valid_5 = accuracy_score(target_valid,predict_forest_valid_5)

predict_forest_test_5 = model_forest_5.predict(feature_test)
result_forest_test_5 = accuracy_score(target_test,predict_forest_test_5)

print('-----------------------------------------------------------')
print (f'Hasil akurasi validation = {result_forest_valid_5*100}, Hasil akurasi test = {result_forest_test_5*100}')

-----------------------------------------------------------
Hasil akurasi validation = 76.82737169517885, Hasil akurasi test = 78.0715396578538


tuning Bootstrap = False, membuat hasil valid dan test menurun, kita akan buat default

In [ ]:
#dikarenakan default oob_score adalah False,
model_forest_6 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',oob_score=True)
model_forest_6.fit(feature_train,target_train)

predict_forest_valid_6 = model_forest_6.predict(feature_valid)
result_forest_valid_6 = accuracy_score(target_valid,predict_forest_valid_6)

predict_forest_test_6 = model_forest_6.predict(feature_test)
result_forest_test_6 = accuracy_score(target_test,predict_forest_test_6)

print('-----------------------------------------------------------')
print (f'Hasil akurasi validation = {result_forest_valid_6*100}, Hasil akurasi test = {result_forest_test_6*100}')

-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625


tuning oob_score=True tidak membuat hasil valid dan test berubah, kita akan buat default

##### warm_start

In [ ]:
#dikarenakan default warm_start adalah False,
model_forest_7 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',warm_start=True)
model_forest_7.fit(feature_train,target_train)

predict_forest_valid_7 = model_forest_7.predict(feature_valid)
result_forest_valid_7 = accuracy_score(target_valid,predict_forest_valid_7)

predict_forest_test_7 = model_forest_7.predict(feature_test)
result_forest_test_7 = accuracy_score(target_test,predict_forest_test_7)

print('-----------------------------------------------------------')
print (f'Hasil akurasi validation = {result_forest_valid_7*100}, Hasil akurasi test = {result_forest_test_7*100}')

-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625


tuning warm_start=True tidak membuat hasil valid dan test berubah, kita akan buat default

##### class_weight

In [ ]:
class_weight = ['balanced', 'balanced_subsample',None]
for weight in class_weight:
    model_forest_8 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy',class_weight=weight)
    model_forest_8.fit(feature_train,target_train)

    predict_forest_valid_8 = model_forest_8.predict(feature_valid)
    result_forest_valid_8 = accuracy_score(target_valid,predict_forest_valid_8)

    predict_forest_test_8 = model_forest_8.predict(feature_test)
    result_forest_test_8 = accuracy_score(target_test,predict_forest_test_8)

    print('-----------------------------------------------------------')
    print (f'Hasil akurasi validation = {result_forest_valid_8*100},Hasil akurasi test = {result_forest_test_8*100} class_weight={weight} ')

-----------------------------------------------------------
Hasil akurasi validation = 79.16018662519441,Hasil akurasi test = 79.16018662519441 class_weight=balanced 
-----------------------------------------------------------
Hasil akurasi validation = 79.47122861586314,Hasil akurasi test = 79.31570762052877 class_weight=balanced_subsample 
-----------------------------------------------------------
Hasil akurasi validation = 79.31570762052877,Hasil akurasi test = 79.93779160186625 class_weight=None 


tuning class_weight=None lebih baik dari yang balanced dan balanced_subsample

#### Kesimpulan

model RandomForestClassifier dengan tuning Hyperparameter, random_state=12345, n_estimators =200, criterion=entropy
memiliki hasil paling bagus dari yang lain nya

In [ ]:
model_forest_9 = RandomForestClassifier(random_state=12345,n_estimators=200,criterion='entropy')
model_forest_9.fit(feature_train,target_train)

predict_forest_valid_9 = model_forest_9.predict(feature_valid)
result_forest_valid_9 = accuracy_score(target_valid,predict_forest_valid_9)

predict_forest_test_9 = model_forest_9.predict(feature_test)
result_forest_test_9 = accuracy_score(target_test,predict_forest_test_9)

print (f'Hasil akurasi validation = {result_forest_valid_9*100}, Hasil akurasi test = {result_forest_test_9*100}')

Hasil akurasi validation = 79.31570762052877, Hasil akurasi test = 79.93779160186625


### Logistic Regression

#### Tanpa Tuning Hyperparameter

In [ ]:
model_log = LogisticRegression(random_state=12345)
model_log.fit(feature_train,target_train)

predict_log_train = model_log.predict(feature_train)
result_log_train = accuracy_score(target_train,predict_log_train)

predict_log_valid = model_log.predict(feature_valid)
result_log_valid = accuracy_score(target_valid,predict_log_valid)

print (f'Hasil akurasi train = {result_log_train*100}, Hasil akurasi validation = {result_log_valid*100}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263


dari hasil model tanpa tuning sudah cukup bagus diatas 75%

#### Tuning Hyperparameter

##### solver

In [ ]:
solver = ['lbfgs', 'liblinear', 'newton-cg', 'sag', 'saga']
for sol in solver:
    model_log_1 = LogisticRegression(random_state=12345,solver=sol)
    model_log_1.fit(feature_train,target_train)

    predict_log_train_1 = model_log_1.predict(feature_train)
    result_log_train_1 = accuracy_score(target_train,predict_log_train_1)

    predict_log_valid_1 = model_log_1.predict(feature_valid)
    result_log_valid_1 = accuracy_score(target_valid,predict_log_valid_1)

    print (f'Hasil akurasi train = {result_log_train_1*100}, Hasil akurasi validation = {result_log_valid_1*100}, solver ={sol}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263, solver =lbfgs
Hasil akurasi train = 74.22199170124482, Hasil akurasi validation = 75.11664074650078, solver =liblinear
Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263, solver =newton-cg
Hasil akurasi train = 69.76141078838174, Hasil akurasi validation = 69.67340590979782, solver =sag
Hasil akurasi train = 69.76141078838174, Hasil akurasi validation = 69.67340590979782, solver =saga


/opt/conda/lib/python3.9/site-packages/scipy/optimize/_linesearch.py:456: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/conda/lib/python3.9/site-packages/scipy/optimize/_linesearch.py:305: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "
/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "


tuning solver=newton-cg dan lbfgs membuat akurasi meningkat, dan tuning sag/saga malah membuat akurasi menurun

##### penalty

In [ ]:
penalty = ['l1','l2']
for penal in penalty:
    model_log_2 = LogisticRegression(random_state=12345,solver='liblinear' ,penalty=penal)
    model_log_2.fit(feature_train,target_train)

    predict_log_train_2 = model_log_2.predict(feature_train)
    result_log_train_2 = accuracy_score(target_train,predict_log_train_2)

    predict_log_valid_2 = model_log_2.predict(feature_valid)
    result_log_valid_2 = accuracy_score(target_valid,predict_log_valid_2)

    print (f'Hasil akurasi train = {result_log_train_2*100}, Hasil akurasi validation = {result_log_valid_2*100}, penalty ={penal}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263, penalty =l1
Hasil akurasi train = 74.22199170124482, Hasil akurasi validation = 75.11664074650078, penalty =l2


dari hasil tuning penalty di dapat
solver = liblinear hanya bisa menampung penalty= l1 dan l2
solver = lbfgs dan newton-cg hanya bisa menampung penalty = l2 dan 'none'
solver = sag/saga hanya bisa menampung penalty= elasticnet

hasil solver liblinear dan penalty l1 akan menjadi pilihan untuk diteruskan tuning

##### multi_class

In [ ]:
multi_class = ['auto', 'ovr']
for multi in multi_class:
    model_log_3 = LogisticRegression(random_state=12345,solver='liblinear' ,penalty='l1',multi_class=multi)
    model_log_3.fit(feature_train,target_train)

    predict_log_train_3 = model_log_3.predict(feature_train)
    result_log_train_3 = accuracy_score(target_train,predict_log_train_3)

    predict_log_valid_3 = model_log_3.predict(feature_valid)
    result_log_valid_3 = accuracy_score(target_valid,predict_log_valid_3)

    print (f'Hasil akurasi train = {result_log_train_3*100}, Hasil akurasi validation = {result_log_valid_3*100}, multi_class ={multi}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263, multi_class =auto
Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263, multi_class =ovr


setelah tuning multi_class dilakukan tidak ada perubahan pada akurasi, dan multi_class=multinomial, tidak bisa di pasangkan dengan solver=liblinear

##### warm_start

In [ ]:
model_log_4 = LogisticRegression(random_state=12345,solver='liblinear' ,penalty='l1',warm_start=True)
model_log_4.fit(feature_train,target_train)

predict_log_train_4 = model_log_4.predict(feature_train)
result_log_train_4 = accuracy_score(target_train,predict_log_train_4)

predict_log_valid_4 = model_log_4.predict(feature_valid)
result_log_valid_4 = accuracy_score(target_valid,predict_log_valid_4)

print (f'Hasil akurasi train = {result_log_train_4*100}, Hasil akurasi validation = {result_log_valid_4*100}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263


setelah tuning warm_start=True tidak merubah hasil akurasi dari train dan validation,

#### Kesimpulan
model LogisticRegression dengan tuning Hyperparameter, random_state=12345, solver=liblinear, penalty=l1
memiliki hasil paling bagus dari yang lain nya


In [ ]:
model_log_5 = LogisticRegression(random_state=12345,solver='liblinear' ,penalty='l1')
model_log_5.fit(feature_train,target_train)

predict_log_train_5 = model_log_5.predict(feature_train)
result_log_train_5 = accuracy_score(target_train,predict_log_train_5)

predict_log_valid_5 = model_log_5.predict(feature_valid)
result_log_valid_5 = accuracy_score(target_valid,predict_log_valid_5)

predict_log_test_5 = model_log_5.predict(feature_test)
result_log_test_5 = accuracy_score(target_test,predict_log_test_5)

print (f'Hasil akurasi train = {result_log_train_5*100}, Hasil akurasi validation = {result_log_valid_5*100}')

Hasil akurasi train = 75.10373443983403, Hasil akurasi validation = 75.89424572317263


## Kesimpulan

Diawal kita melakukan sanity check data manual antara data 'is_ultra' 0=69% dan 1=31%, yang mengharuskan model memiliki akurasi lebih dari 75%, dengan menggunakan :
1. DecisionTreeClassifier dengan tuning hyperparameter random_state=12345, max_depth =14, criterion=gini, splitter=best, menghasilkan model yang tidak underfitting maupun overfitting dengan tingkat akurasi pada test sebesar 75.73872472783826%

2. RandomForestClassifier dengan tuning hyperparameter random_state=12345, n_estimators =200, criterion=entropy, menghasilkan model yang tidak underfitting maupun ovefitting dengan tingkat akurasi pada test sebesar 79.93779160186625

3. Logistic Regression dengan tuning random_state=12345, solver=liblinear, penalty=l1, menghasilkan model yang tidak underfitting maupun overfitting dengan tingkat akurasi pada validation sebesar 75.89424572317263

dari ketiga model di atas, RandomForestClassifier memiliki tingkat akurasi tertinggi, tetapi jika dibandingkan kecepetan eksekusi program sedikit lebih lambat dari model yang lain